Split cardboard up across two tables part deux

mcwhittemore opened the pull request:
Closes #156
Closes #184
Closes #149
Closes #145

This grows off of [the ice pr](https://github.com/mapbox/cardboard/pull/185) and also removes `cardboard.list`.

It also removes `cardboard.batch.*` and converts `put`, `get`, and `del` on the main `cardboard` object to be batch first apis.

/cc @rclark & @jakepruitt 

**Notes from #185**

The goal of this PR is to resolve hot partition problems as outlined in https://github.com/mapbox/cardboard/issues/184.

**TODO**

- Make sure we are only making calls to search when a feature is created/deleted
- Add batch back
- Fix inline docs
- Update api.md
- Drop test support for older versions of node
- Add test support for newer versions of node
- convert tests to tape

**Things changed**

- Moved to `geobuf@3.0.0` which requires node@4.5.0
- changed `createTable` to `createTables` and removed name overrides.
- changed `list` with no callback to `listStream`. If we want to keep the old functionality, we can have `list` call `listStream`, I just wanted to clean up the code paths in that function.
- no mas bbox
- changed `batch.remove` to `batch.del` to match the non-batch api

mcwhittemore commented:
List work as begun to happen in the [cardboar-list](https://github.com/mapbox/cardboard-list) repo.

rclark commented on .eslintrc:
```diff
@@ -5,7 +5,8 @@
     "env": {
-        "node": true
+        "node": true,
+        "mocha": true
```
What does this do?

mcwhittemore replied:
lets eslint lint mocha tests

rclark commented on bin/cardboard.js:
```diff
@@ -81,20 +66,13 @@ cardboard.createTable(function(err){
 
     if (command === 'list') {
-        return cardboard.list(dataset)
+        return cardboard.listStream(dataset)
```
Shouldn't this command block be entirely removed?

rclark commented on lib/main_table.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "AttributeDefinitions": [
+        {"AttributeName": "index", "AttributeType": "S"}
```
Could we call this `featureid` or something with `id` in it?

mcwhittemore replied:
yea, my problem with id is that this string is made of two other things called id so it gets confusing to keep calling things id.

What about `key`?

rclark replied:
key sounds perfect. I wrote this before I read further down and saw the compoundness.

rclark commented on lib/metadata.js:
```diff
@@ -1,395 +0,0 @@
-var queue = require('queue-async');
```
Where are you imagining the metadata code is going to live in the end?

rclark commented on lib/utils.js:
```diff
@@ -30,7 +30,7 @@ function Utils(config) {
                 if (val) {
                     try {
-                        feature = geobuf.geobufToFeature(val);
+                        feature = geobuf.decode(new Pbf(val));
```
We should make sure we have a plan for what the migrations and uptime are going to look like before we mix the geobuf upgrade into this.

rclark commented on lib/utils.js:
```diff
@@ -60,6 +60,49 @@ function Utils(config) {
+        var q = queue(100);
+
+        featureIds.forEach(function (id) {
+            // TODO: have this use bulk gets
```
This is TODO in this PR, or after?

rclark commented on lib/utils.js:
```diff
@@ -91,31 +134,31 @@ function Utils(config) {
+            north: bounds[3]
+        };
+        var buf = utils.encodeFeature(f);
         var useS3 = buf.length >= config.MAX_GEOMETRY_SIZE;
```
Given that we don't utilize the S3 side of this as was envisioned originally, and given that this is such a breaking PR anyway, we should consider removing this stuff entirely.

rclark commented on lib/utils.js:
```diff
@@ -91,31 +134,31 @@ function Utils(config) {
+        var databaseFeature = utils.createFeatureKey(dataset, primary);
+        databaseFeature.val = useS3 ? null : buf;
+        databaseFeature.size = buf.length;
+        databaseFeature.west = truncateNum(info.west);
```
Since we don't have a BBOX query here anymore (moving to another module / index in ultimately), is there any reason for our main table features to contain any more information than id and geobuf?

rclark commented on package.json:
```diff
@@ -5,8 +5,8 @@
     "pretest": "eslint index.js lib test/*.js",
-    "test": "nyc tape test/*.test.js",
-    "coverage": "nyc --reporter html tape test/*.test.js && opener ./coverage/index.html",
+    "test": "nyc mocha test/*.test.js",
```
:-1: Not into this.

mcwhittemore replied:
moving from tape to mocha?

I find it almost impossible to work with tape when tests are failing which is 90% of a refactor. The output is really hard to read and I get rather confused about what is happening. Mocha also exits the same way so no matter what caused the test to fail (a bad assert or invalid syntax due to an unexpected result) you see the same report.

rclark replied:
If you want better output put `tap-spec` in as a dev dependency and make the test script pipe into it:

```
"test": "nyc tape test/*.test.js | tap-spec
```

We need to maintain test consistency in our backend systems to maintain human sanity as we move from one project to another and another and back again. There's no doubt that the test suite in cardboard is crummy, but that's not tape's fault. 

You're removing so much code that I almost wonder if the test suite is salvageable. Sometimes these heavy refactors call for a test suite rewrite, where you just recreate the old assertions in a more-sane fashion.

rclark commented on lib/utils.js:
```diff
@@ -125,8 +168,9 @@ function Utils(config) {
+        bits.shift();
         bits.shift();
         return bits.join('!');
     };
```
Given how much simplification has happened to the routines in `utils.js`, it might not make sense to make a separate module for them anymore. Are they even used in more than one place?

mcwhittemore replied:
They are used in lots of tests.

mcwhittemore commented:
@rclark - I think this is ready for one last review.

I decided to keep utils in cardboard for now. No need to add repos.

Big changes since you last looked.

- removed `feature` from the key. This was here to stop collisions in a multi-index table. As we've decided this isn't how we want cardboard to be implemented, removing `features` should help our hash be partitioned well.
- added `Cardboard.streamHelper` - this takes an array of action types and recordsHandler and returns a function that can handle a dynamodb stream event in a lambda context. The recordsHandler will be passed an array of records and a callback if there are records that match the action filter.

mcwhittemore commented:
One leftover question is if cardboard should really do `batchWriteAll` for all of its writes.
